download can silently fail?

As reported by @yuhaia on MacOS Catalina 10.15.7 with Python 3.7 for `macavaney:anserini-trec-robust04`'s anserini index download. https://github.com/allenai/ir_datasets/issues/52#issuecomment-814020499

The best place to start would be to simulate a failed download on the same OS, python version, and dataset and see if I can reproduce it. If not, I'm not sure what to try next.

@yuhaia-- if there are any more details on this that could help me get to the bottom of this problem, please let me know!

```bash
Successfully built ir-datasets warc3-wet-clueweb09 unlzw3 cbor
Installing collected packages: cbor, trec-car-tools, lz4, warc3-wet, warc3-wet-clueweb09, zlib-state, ijson, pyautocorpus, unlzw3, ir-datasets
Successfully installed cbor-1.0.0 ijson-3.1.4 ir-datasets-0.5.0 lz4-3.1.10 pyautocorpus-0.1.6 trec-car-tools-2.5.4 unlzw3-0.2.1 warc3-wet-0.2.3 warc3-wet-clueweb09-0.2.5 zlib-state-0.1.3
WARNING: You are using pip version 20.2.4; however, version 21.3.1 is available.
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.
root@f535407a1f28:/ir_datasets# python -m test.integration.trec_robust04
trec-robust04 docs: 0doc [01:17, ?doc/s]
[INFO] [finished] trec-robust04 docs: [01:17] [0doc] [?doc/s]
[INFO] [starting] doc lookups by index
[INFO] [starting] building docstore
docs_iter:   0%|                                    | 0/528155 [01:15<?, ?doc/s]
[INFO] [finished] docs_iter: [01:15] [0doc] [?doc/s]
[INFO] [finished] building docstore [01:16]
[INFO] [error] doc lookups by index [01:16]
[INFO] [starting] doc lookups by doc_id
[INFO] [starting] building docstore
docs_iter:   0%|                                    | 0/528155 [01:17<?, ?doc/s]
[INFO] [finished] docs_iter: [01:17] [0doc] [?doc/s]
[INFO] [finished] building docstore [01:17]
[INFO] [error] doc lookups by doc_id [01:17]
[INFO] Please confirm you agree to the TREC Robust 2004 data usage agreement found at <https://trec.nist.gov/data/cd45/index.html>
[INFO] If you have a local copy of https://trec.nist.gov/data/robust/qrels.robust2004.txt, you can symlink it here to avoid downloading it again: /tf/data/external/IR_DATASETS/downloads/123c2a0ba2ec31178cb1050995dcfdfa
[INFO] [starting] https://trec.nist.gov/data/robust/qrels.robust2004.txt
[INFO] [finished] https://trec.nist.gov/data/robust/qrels.robust2004.txt: [00:01] [6.54MB] [5.77MB/s]
trec-robust04 qrels: 311410qrel [00:03, 96275.73qrel/s]                                           
[INFO] [finished] trec-robust04 qrels: [00:03] [311410qrel] [96274.20qrel/s]       
trec-robust04/fold1 qrels: 62789qrel [00:01, 57460.09qrel/s]
[INFO] [finished] trec-robust04/fold1 qrels: [00:01] [62789qrel] [57457.44qrel/s]
trec-robust04/fold2 qrels: 63917qrel [00:01, 57797.58qrel/s]
[INFO] [finished] trec-robust04/fold2 qrels: [00:01] [63917qrel] [57794.48qrel/s]
trec-robust04/fold3 qrels: 62901qrel [00:01, 57734.98qrel/s]
[INFO] [finished] trec-robust04/fold3 qrels: [00:01] [62901qrel] [57732.31qrel/s]
trec-robust04/fold4 qrels: 57962qrel [00:01, 52777.53qrel/s]
[INFO] [finished] trec-robust04/fold4 qrels: [00:01] [57962qrel] [52774.37qrel/s]
trec-robust04/fold5 qrels: 63841qrel [00:01, 58219.68qrel/s]
[INFO] [finished] trec-robust04/fold5 qrels: [00:01] [63841qrel] [58217.09qrel/s]
[INFO] [starting] https://trec.nist.gov/data/robust/04.testset.gz
[INFO] [finished] https://trec.nist.gov/data/robust/04.testset.gz: [00:00] [34.3kB] [309kB/s]
trec-robust04 queries: 250query [00:00, 395.46query/s]                                     
[INFO] [finished] trec-robust04 queries: [00:00] [250query] [395.42query/s]
trec-robust04/fold1 queries: 50query [00:00, 1514.23query/s]
[INFO] [finished] trec-robust04/fold1 queries: [00:00] [50query] [1511.82query/s]
trec-robust04/fold2 queries: 50query [00:00, 1509.87query/s]
[INFO] [finished] trec-robust04/fold2 queries: [00:00] [50query] [1507.54query/s]
trec-robust04/fold3 queries: 50query [00:00, 1508.60query/s]
[INFO] [finished] trec-robust04/fold3 queries: [00:00] [50query] [1506.40query/s]
trec-robust04/fold4 queries: 50query [00:00, 1504.80query/s]
[INFO] [finished] trec-robust04/fold4 queries: [00:00] [50query] [1502.66query/s]
trec-robust04/fold5 queries: 50query [00:00, 1503.55query/s]
[INFO] [finished] trec-robust04/fold5 queries: [00:00] [50query] [1501.43query/s]
.
======================================================================
ERROR: test_trec_robust04_docs (__main__.TestTrecRobust04) [docs_iter split] (dataset='trec-robust04')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/ir_datasets/test/integration/base.py", line 38, in _test_docs
    self._assert_namedtuple(next(it[idx:idx+1]), doc)
  File "/ir_datasets/ir_datasets/indices/lz4_pickle.py", line 54, in __next__
    raise StopIteration
StopIteration

======================================================================
ERROR: test_trec_robust04_docs (__main__.TestTrecRobust04) [docs_store] (dataset='trec-robust04')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/ir_datasets/test/integration/base.py", line 46, in _test_docs
    ret_doc = doc_store.get(doc.doc_id)
  File "/ir_datasets/ir_datasets/indices/base.py", line 13, in get
    raise KeyError(f'doc_id={doc_id} not found')
KeyError: 'doc_id=FBIS3-1 not found'

======================================================================
FAIL: test_trec_robust04_docs (__main__.TestTrecRobust04) [docs] (dataset='trec-robust04')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/ir_datasets/test/integration/base.py", line 29, in _test_docs
    self.assertEqual(expected_count, count)
AssertionError: 528155 != 0

----------------------------------------------------------------------
Ran 3 tests in 239.856s

FAILED (failures=1, errors=2)
```

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -14,7 +14,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8, 3.9]
+        python-version: [3.7, 3.8, 3.9, '3.10']
         os: ['ubuntu-latest', 'windows-latest', 'macOs-latest']
         architecture: ['x64']
 

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -14,7 +14,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8, 3.9, '3.10']
+        python-version: [3.7, 3.8, 3.9]
         os: ['ubuntu-latest', 'windows-latest', 'macOs-latest']
         architecture: ['x64']
 

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -14,7 +14,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: [3.7, 3.8, 3.9]
         os: ['ubuntu-latest', 'windows-latest', 'macOs-latest']
         architecture: ['x64']
 

--- a/ir_datasets/formats/trec.py
+++ b/ir_datasets/formats/trec.py
@@ -113,8 +113,14 @@ class TrecDocs(BaseDocs):
 
     def _docs_iter(self, path):
         if Path(path).is_file():
-            if str(path).endswith('.gz'):
+            spath = str(path)
+            if spath.endswith('.gz'):
                 with gzip.open(path, 'rb') as f:
+                    yield from self._parser(f)
+            elif any(spath.endswith(ext) for ext in ['.z', '.0z', '.1z', '.2z']):
+                # unix "compress" command encoding
+                unlzw3 = ir_datasets.lazy_libs.unlzw3()
+                with io.BytesIO(unlzw3.unlzw(path)) as f:
                     yield from self._parser(f)
             else:
                 with path.open('rb') as f:

--- a/ir_datasets/formats/trec.py
+++ b/ir_datasets/formats/trec.py
@@ -113,11 +113,11 @@ class TrecDocs(BaseDocs):
 
     def _docs_iter(self, path):
         if Path(path).is_file():
-            spath = str(path)
-            if spath.endswith('.gz'):
+            path_suffix = Path(path).suffix.lower()
+            if path_suffix == '.gz':
                 with gzip.open(path, 'rb') as f:
                     yield from self._parser(f)
-            elif any(spath.endswith(ext) for ext in ['.z', '.0z', '.1z', '.2z']):
+            elif path_suffix in ['.z', '.0z', '.1z', '.2z']:
                 # unix "compress" command encoding
                 unlzw3 = ir_datasets.lazy_libs.unlzw3()
                 with io.BytesIO(unlzw3.unlzw(path)) as f:

--- a/ir_datasets/lazy_libs.py
+++ b/ir_datasets/lazy_libs.py
@@ -102,4 +102,10 @@ def pyautocorpus():
     if 'pyautocorpus' not in _cache:
         import pyautocorpus
         _cache['pyautocorpus'] = pyautocorpus
-    return _cache['pyautocorpus']    
+    return _cache['pyautocorpus']
+
+def unlzw3():
+    if 'unlzw3' not in _cache:
+        import unlzw3
+        _cache['unlzw3'] = unlzw3
+    return _cache['unlzw3']

--- a/requirements.txt
+++ b/requirements.txt
@@ -11,3 +11,4 @@ warc3-wet-clueweb09>=0.2.5
 zlib-state>=0.1.3
 ijson>=3.1.3
 pyautocorpus>=0.1.1
+unlzw3>=0.2.1

